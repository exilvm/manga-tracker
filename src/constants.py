# id for the group named "No group"
NO_GROUP = 1
