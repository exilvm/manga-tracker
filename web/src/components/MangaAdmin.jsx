import { useSnackbar } from 'notistack';
import React, { useCallback, useMemo, useState } from 'react';
import {
  Container,
  Grid,
  Paper,
  Tooltip,
  Typography,
} from '@material-ui/core';

import { makeStyles } from '@material-ui/core/styles';

import {
  Select,
} from 'mui-rff';

import {
  defaultDateDistanceToNow,
  defaultDateFormat,
} from '../utils/utilities';
import {
  AddRowFormTemplate,
  EditableSelect,
  MaterialTable,
} from './MaterialTable';


const useStyles = makeStyles((theme) => ({
  title: {
    width: '75%',
    textAlign: 'left',
    paddingBottom: '10px',
    [theme.breakpoints.down('sm')]: {
      width: '100%',
    },
  },
  titleBar: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  thumbnail: {
    maxWidth: '250px',
    maxHeight: '355px',
    [theme.breakpoints.down('sm')]: {
      maxWidth: '180px',
    },
    [theme.breakpoints.down('xs')]: {
      maxWidth: '125px',
    },
  },
  details: {
    display: 'flex',
  },
  detailText: {
    marginLeft: '5px',
    [theme.breakpoints.down('sm')]: {
      marginLeft: '3px',
    },
  },
  infoTable: {
    marginLeft: '30px',
    marginTop: '3px',
    [theme.breakpoints.down('sm')]: {
      marginLeft: '20px',
    },
    [theme.breakpoints.down('xs')]: {
      marginLeft: '10px',
    },
  },
  paper: {
    padding: '1em',
    minWidth: '440px',
  },
  addRowForm: {
    minWidth: '150px',
  },
}));


function MangaAdmin(props) {
  const {
    mangaData,
  } = props;

  // Constants
  const mangaId = mangaData.manga_id;
  const latestRelease = mangaData.latest_release ?
    new Date(mangaData.latest_release) :
    null;
  const estimatedRelease = new Date(mangaData.estimated_release);

  // Simple hooks
  const classes = useStyles();
  const { enqueueSnackbar } = useSnackbar();
  const [loading, setLoading] = useState(false);
  const [scheduledUpdates, setScheduledUpdates] = useState([]);

  const formatScheduledRuns = useCallback((runs) => runs.map(run => {
    const found = mangaData.services.find(s => s.service_id === run.service_id);
    if (!found) {
      return run;
    }
    return {
      ...run,
      name: found.name,
    };
  }), [mangaData.services]);

  // Data fetching callbacks
  const fetchData = useCallback(() => {
    setLoading(true);

    return fetch(`/api/admin/manga/${mangaId}/scheduledRuns`)
      .then(res => res.json())
      .then(json => {
        setScheduledUpdates(formatScheduledRuns(json.data || []));
      })
      .catch(console.error)
      .finally(() => setLoading(false));
  }, [formatScheduledRuns, mangaId]);

  const onCreateRow = useCallback((form) => {
    fetch(`/api/admin/manga/${mangaId}/scheduledRun/${form.service_id}`, {
      method: 'POST',
    })
      .then(res => {
        if (res.status !== 200) {
          throw new Error(`Failed to create row ${res.status} ${res.statusText}`);
        }
        return res.json();
      })
      .then(json => {
        setScheduledUpdates(formatScheduledRuns([...scheduledUpdates, json.inserted]));
        enqueueSnackbar(
          `Successfully scheduled manga ${mangaId} to be checked on service ${form.service_id}`,
          { variant: 'success' }
        );
      })
      .catch(err => enqueueSnackbar(err.message, { variant: 'error' }));
  }, [mangaId, formatScheduledRuns, scheduledUpdates, enqueueSnackbar]);

  const onDeleteRow = useCallback((row) => {
    const serviceId = row.values.service_id;
    fetch(`/api/admin/manga/${mangaId}/scheduledRun/${serviceId}`, {
      method: 'DELETE',
    })
      .then(res => {
        if (res.status !== 200) {
          enqueueSnackbar(`Failed to delete row. ${res.status} ${res.statusText}`, { variant: 'error' });
        } else {
          setScheduledUpdates(
            formatScheduledRuns(scheduledUpdates.filter(r => r.service_id !== serviceId))
          );
          enqueueSnackbar(
            `Successfully deleted service ${row.values.name} from scheduled runs`,
            { variant: 'success' }
          );
        }
      })
      .catch(err => enqueueSnackbar(err.message, { variant: 'error' }));
  }, [enqueueSnackbar, formatScheduledRuns, mangaId, scheduledUpdates]);

  // Table layout
  const fields = useMemo(() => [
    <Select
      name='service_id'
      data={mangaData.services.map(s => ({ value: s.service_id, label: s.name }))}
      required
    />,
  ], [mangaData.services]);

  const CreateDialog = useMemo(() => ({ open, onClose }) => (
    <AddRowFormTemplate
      fields={fields}
      onSubmit={onCreateRow}
      onClose={onClose}
      open={open}
      formClass={classes.addRowForm}
    />
  ), [fields, onCreateRow, classes.addRowForm]);

  const columns = useMemo(() => [
    {
      Header: 'Service name',
      accessor: 'name',
      EditCell: ({ row, cell, state }) => (
        <EditableSelect
          value={row.values.service_id}
          items={mangaData.services.map(s => ({ value: s.service_id, text: s.name }))}
          cell={cell}
          row={row}
          state={state}
          onChange={(serviceId) => {
            state.rowEditStates[row.id].service_id = serviceId;
          }}
        />
      ),
    },
    { Header: 'Service id', accessor: 'service_id', canEdit: false },
  ], [mangaData.services]);

  return (
    <Container maxWidth='lg' disableGutters>
      <Paper className={classes.paper}>
        <div className={classes.titleBar}>
          <Typography className={classes.title} variant='h4'>{mangaData.title}</Typography>
        </div>
        <div className={classes.details}>
          <a href={mangaData.mal} target='_blank' rel='noreferrer noopener'>
            <img
              src={mangaData.cover}
              className={classes.thumbnail}
              alt={mangaData.title}
            />
          </a>
          <Grid
            container
            justify='space-between'
          >
            <table className={classes.infoTable}>
              <tbody>
                <tr>
                  <td><Typography>Latest release:</Typography></td>
                  <td>
                    <Tooltip title={latestRelease ? latestRelease.toUTCString() : 'Unknown'}>
                      <Typography className={classes.detailText}>
                        {latestRelease ?
                          defaultDateFormat(latestRelease) + ' - ' + defaultDateDistanceToNow(latestRelease) :
                          'Unknown'}
                      </Typography>
                    </Tooltip>
                  </td>
                </tr>
                <tr>
                  <td><Typography>Estimated release interval:</Typography></td>
                  <td>
                    <Typography className={classes.detailText}>
                      {(mangaData.release_interval ?
                        `${mangaData.release_interval?.days || 0} days ${mangaData.release_interval?.hours || 0} hours` :
                        'Unknown')}
                    </Typography>
                  </td>
                </tr>
                <tr>
                  <td><Typography>Estimated next release:</Typography></td>
                  <td>
                    <Typography className={classes.detailText}>
                      {defaultDateFormat(estimatedRelease)}
                    </Typography>
                  </td>
                </tr>
                <tr>
                  <td><Typography>Latest chapter:</Typography></td>
                  <td>
                    <Typography className={classes.detailText}>
                      {mangaData.latest_chapter ? mangaData.latest_chapter : 'Unknown'}
                    </Typography>
                  </td>
                </tr>
              </tbody>
            </table>
          </Grid>
        </div>
        <MaterialTable
          data={scheduledUpdates}
          columns={columns}
          editable
          deletable
          creatable
          CreateDialog={CreateDialog}
          title='Forced checks'
          fetchData={fetchData}
          onDeleteRow={onDeleteRow}
          loading={loading}
        />
      </Paper>
    </Container>
  );
}

export default MangaAdmin;
