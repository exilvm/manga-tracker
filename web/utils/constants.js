const csrfMissing = 'Missing CSRF token';
const isDev = process.env.NODE_ENV !== 'production';
const isTest = process.env.NODE_ENV === 'test';


module.exports = {
  csrfMissing,
  isDev,
  isTest,
};
