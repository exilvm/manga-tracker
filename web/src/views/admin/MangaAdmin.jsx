import {
  Container,
  Grid,
  IconButton,
  Paper,
  Tooltip,
  Typography,
} from '@material-ui/core';

import { makeStyles } from '@material-ui/core/styles';

import { SubdirectoryArrowLeft as SubdirectoryArrowLeftIcon } from '@material-ui/icons';
import { useConfirm } from 'material-ui-confirm';
import PropTypes from 'prop-types';

import { Select } from 'mui-rff';
import Link from 'next/link';
import { useSnackbar } from 'notistack';
import React, { useCallback, useMemo, useState } from 'react';
import MangaAliases from '../../components/MangaAliases';
import MangaInfo from '../../components/MangaInfo';


import {
  AddRowFormTemplate,
  EditableSelect,
  MaterialTable,
} from '../../components/MaterialTable';
import { useCSRF } from '../../utils/csrf';
import { getManga } from '../../api/manga';
import {
  getScheduledRuns,
  createScheduledRun,
  deleteScheduledRun,
} from '../../api/admin/manga';
import { MangaCover } from '../../components/MangaCover';


const useStyles = makeStyles((theme) => ({
  title: {
    width: '75%',
    textAlign: 'left',
    paddingBottom: '10px',
    [theme.breakpoints.down('sm')]: {
      width: '100%',
    },
  },
  titleBar: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  details: {
    display: 'flex',
    flexFlow: 'row',
    [theme.breakpoints.down('xs')]: {
      flexFlow: 'wrap',
      justifyContent: 'center',
    },
  },
  paper: {
    padding: '1em',
    minWidth: '400px',
  },
  addRowForm: {
    minWidth: '150px',
  },
  infoGrid: {
    marginLeft: theme.spacing(4),
    width: 'fit-content',
    [theme.breakpoints.down('xs')]: {
      marginLeft: '0px',
    },
  },
}));


function MangaAdmin(props) {
  const {
    mangaData: {
      manga,
      services,
      aliases: aliasesProp,
    },
    serviceConfigs,
  } = props;

  // Constants
  const mangaId = manga.mangaId;

  // Hooks
  const classes = useStyles();
  const { enqueueSnackbar } = useSnackbar();
  const csrf = useCSRF();
  const confirm = useConfirm();

  const [loading, setLoading] = useState(false);
  const [scheduledUpdates, setScheduledUpdates] = useState([]);
  const [aliases, setAliases] = useState(aliasesProp);
  const [mangaTitle, setMangaTitle] = useState(manga.title);

  const formatScheduledRuns = useCallback((runs) => runs.map(run => {
    const found = services.find(s => s.serviceId === run.serviceId);
    if (!found) {
      return run;
    }
    return {
      ...run,
      name: found.name,
    };
  }), [services]);

  const onTitleChange = useCallback(() => {
    getManga(mangaId)
      .then(data => {
        setAliases(data.aliases);
        setMangaTitle(data.manga.title);
      });
  }, [mangaId]);

  // Data fetching callbacks
  const fetchData = useCallback(() => {
    setLoading(true);

    return getScheduledRuns(mangaId)
      .then(data => {
        setScheduledUpdates(formatScheduledRuns(data || []));
      })
      .finally(() => setLoading(false));
  }, [formatScheduledRuns, mangaId]);

  const onCreateRow = useCallback((form) => {
    createScheduledRun(csrf, mangaId, form.serviceId)
      .then(json => {
        setScheduledUpdates(formatScheduledRuns([...scheduledUpdates, json.inserted]));
        enqueueSnackbar(
          `Successfully scheduled manga ${mangaId} to be checked on service ${form.serviceId}`,
          { variant: 'success' }
        );
      })
      .catch(err => enqueueSnackbar(err.message, { variant: 'error' }));
  }, [mangaId, formatScheduledRuns, scheduledUpdates, enqueueSnackbar, csrf]);

  const onDeleteRow = useCallback((row) => {
    const serviceId = row.values.serviceId;
    deleteScheduledRun(csrf, mangaId, serviceId)
      .then(() => {
        setScheduledUpdates(
          formatScheduledRuns(scheduledUpdates.filter(r => r.serviceId !== serviceId))
        );
        enqueueSnackbar(
          `Successfully deleted service ${row.values.name} from scheduled runs`,
          { variant: 'success' }
        );
      })
      .catch(err => enqueueSnackbar(err.message, { variant: 'error' }));
  }, [enqueueSnackbar, formatScheduledRuns, mangaId, scheduledUpdates, csrf]);

  // Table layout
  const fields = useMemo(() => {
    const servicesWithRunsEnabled = new Set(
      serviceConfigs.filter(s => s.scheduledRunsEnabled).map(s => s.serviceId)
    );
    const data = services
      ?.filter(s => servicesWithRunsEnabled.has(s.serviceId))
      .map(s => ({ value: s.serviceId, label: s.name }));

    return [
      <Select
        name='serviceId'
        key='serviceId'
        label='Service'
        SelectDisplayProps={{ 'aria-label': 'Service select' }}
        data={data}
        required
      />,
    ];
  }, [services, serviceConfigs]);

  const CreateDialog = useMemo(() => ({ open, onClose }) => (
    <AddRowFormTemplate
      fields={fields}
      onSubmit={onCreateRow}
      onClose={onClose}
      open={open}
      formClass={classes.addRowForm}
    />
  ), [fields, onCreateRow, classes.addRowForm]);

  const columns = useMemo(() => [
    {
      Header: 'Service name',
      accessor: 'name',
      EditCell: ({ row, cell, state }) => (
        <EditableSelect
          value={row.values.serviceId}
          items={services.map(s => ({ value: s.serviceId, text: s.name }))}
          cell={cell}
          row={row}
          state={state}
          onChange={(serviceId) => {
            state.rowEditStates[row.id].serviceId = serviceId;
          }}
        />
      ),
    },
    { Header: 'Service id', accessor: 'serviceId', canEdit: false },
  ], [services]);

  return (
    <Container maxWidth='lg' disableGutters>
      <Paper className={classes.paper}>
        <div className={classes.titleBar}>
          <Typography className={classes.title} variant='h4'>{mangaTitle}</Typography>
          <Link href={`/manga/${mangaId}`} passHref>
            <Tooltip title='Go back' aria-label='go back to manga page'>
              <IconButton>
                <SubdirectoryArrowLeftIcon />
              </IconButton>
            </Tooltip>
          </Link>
        </div>
        <div className={classes.details}>
          <a href={manga.mal} target='_blank' rel='noreferrer noopener'>
            <MangaCover
              url={manga.cover}
              alt={manga.title}
            />
          </a>
          <Grid
            container
            direction='column'
            className={classes.infoGrid}
          >
            <MangaInfo mangaData={manga} />
            <MangaAliases
              aliases={aliases}
              mangaId={mangaId}
              onTitleUpdate={onTitleChange}
              enqueueSnackbar={enqueueSnackbar}
              confirm={confirm}
              allowEdits
            />
          </Grid>
        </div>
        <MaterialTable
          data={scheduledUpdates}
          columns={columns}
          editable
          deletable
          creatable
          CreateDialog={CreateDialog}
          title='Scheduled runs'
          fetchData={fetchData}
          onDeleteRow={onDeleteRow}
          loading={loading}
          toolbarProps={{ addButtonLabel: 'add scheduled run' }}
        />
      </Paper>
    </Container>
  );
}

MangaAdmin.propTypes = {
  mangaData: PropTypes.shape({
    manga: PropTypes.object.isRequired,
    services: PropTypes.arrayOf(PropTypes.object),
    aliases: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  serviceConfigs: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default MangaAdmin;
