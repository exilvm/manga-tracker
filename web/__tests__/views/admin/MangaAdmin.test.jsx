import { act, render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import fetchMock from 'fetch-mock';
import React from 'react';

import {
  enqueueSnackbarMock,
  expectErrorSnackbar,
  expectSuccessSnackbar,
  mockNotistackHooks,
  mockUTCDates,
  muiSelectValue,
} from '../../utils';
import MangaAdmin from '../../../src/views/admin/MangaAdmin';
import { fullManga } from '../../constants';


beforeEach(() => mockNotistackHooks());

const mangaId = fullManga.manga.manga_id;

describe('Manga admin page should render correctly', () => {
  mockUTCDates();
  fetchMock.get('*', []);

  it('should render correctly without data', async () => {
    let baseElement;
    await act(async () => {
      ({ baseElement } = render(<MangaAdmin mangaData={{ manga: {}}} />));
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should render correctly with data', async () => {
    let baseElement;
    await act(async () => {
      ({ baseElement } = render(<MangaAdmin mangaData={fullManga} />));
    });

    expect(baseElement).toMatchSnapshot();
  });
});

describe('Manga admin page should handle data fetching correctly', () => {
  const mockServices = fullManga.services.map(s => ({
    service_id: s.service_id,
    manga_id: mangaId,
  }));
  const getMock = jest.fn().mockReturnValue({ data: mockServices });

  beforeEach(() => {
    fetchMock.reset();
    fetchMock.get(`/api/admin/manga/${mangaId}/scheduledRuns`, getMock);
  });

  const renderPage = async () => {
    await act(async () => {
      render(<MangaAdmin mangaData={fullManga} />);
    });
    expect(getMock).toHaveBeenCalledTimes(1);
  };

  const submitForm = async () => {
    await act(async () => {
      userEvent.click(screen.getByRole('button', { name: 'Create row' }));
    });
  };

  const deleteRow = async () => {
    await act(async () => {
      userEvent.click(screen.getByRole('button', { name: 'Confirm delete row' }));
    });
  };

  it('Should fetch services on render', async () => {
    await renderPage();

    fullManga.services.forEach(s => {
      expect(screen.getByText(s.name)).toBeTruthy();
    });
  });

  it('Should call the correct endpoint on adding new run', async () => {
    const serviceId = fullManga.services[0].service_id;
    const serviceName = fullManga.services[0].name;
    const postMock = jest.fn().mockReturnValue({ inserted: { service_id: serviceId, name: serviceName }});
    const partialGetMock = jest.fn().mockReturnValue({ data: mockServices.slice(1, 2) });

    fetchMock.post(`/api/admin/manga/${mangaId}/scheduledRun/${serviceId}`, postMock);
    fetchMock.get(`/api/admin/manga/${mangaId}/scheduledRuns`, partialGetMock, { overwriteRoutes: true });

    await act(async () => {
      render(<MangaAdmin mangaData={fullManga} />);
    });
    expect(partialGetMock).toHaveBeenCalledTimes(1);

    expect(screen.queryByText(serviceName)).toBeNull();

    userEvent.click(screen.getByLabelText('Add item'));

    const form = within(screen.getByRole('presentation', { name: 'Create item form' }));
    muiSelectValue(form, 'Service select', serviceName);

    await submitForm();

    expect(postMock).toHaveBeenCalledTimes(1);
    expect(screen.queryByText(serviceName)).toBeTruthy();
    expectSuccessSnackbar();
  });

  it('Should remove row on delete', async () => {
    const serviceId = fullManga.services[0].service_id;
    const serviceName = fullManga.services[0].name;
    const deleteMock = jest.fn().mockReturnValue({});

    fetchMock.delete(`/api/admin/manga/${mangaId}/scheduledRun/${serviceId}`, deleteMock);

    await renderPage();

    const rowElement = screen.getByText(serviceName).closest('tr');
    const row = within(rowElement);

    userEvent.click(row.getByRole('button', { name: 'Delete row' }));

    await deleteRow();

    expect(deleteMock).toHaveBeenCalledTimes(1);

    expect(screen.queryByText(serviceName)).toBeNull();
    expectSuccessSnackbar();
  });

  it('Should show error snackbar on error', async () => {
    const serviceId = fullManga.services[0].service_id;
    const serviceName = fullManga.services[0].name;

    fetchMock.delete(`/api/admin/manga/${mangaId}/scheduledRun/${serviceId}`, 500);
    fetchMock.post(`/api/admin/manga/${mangaId}/scheduledRun/${serviceId}`, 500);

    await renderPage();

    const rowElement = screen.getByText(serviceName).closest('tr');
    const row = within(rowElement);

    userEvent.click(row.getByRole('button', { name: 'Delete row' }));

    await deleteRow();

    // Should not delete on error
    expect(screen.queryByText(serviceName)).toBeTruthy();
    expect(enqueueSnackbarMock).toHaveBeenCalledTimes(1);
    expectErrorSnackbar();

    userEvent.click(screen.getByLabelText('Add item'));

    const form = within(screen.getByRole('presentation', { name: 'Create item form' }));
    muiSelectValue(form, 'Service select', serviceName);

    await submitForm();
    expect(enqueueSnackbarMock).toHaveBeenCalledTimes(2);
    expectErrorSnackbar();
  });
});
