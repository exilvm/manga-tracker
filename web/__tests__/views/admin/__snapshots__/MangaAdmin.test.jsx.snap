// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manga admin page should render correctly should render correctly with data 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
    >
      <div
        class="MuiPaper-root makeStyles-paper-23 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="makeStyles-titleBar-20"
        >
          <h4
            class="MuiTypography-root makeStyles-title-19 MuiTypography-h4"
          >
            Dr. STONE
          </h4>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            title="Go back"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 9l1.42 1.42L8.83 14H18V4h2v12H8.83l3.59 3.58L11 21l-6-6 6-6z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="makeStyles-details-22"
        >
          <a
            href="https://myanimelist.net/manga/103897"
            rel="noreferrer noopener"
            target="_blank"
          >
            <img
              alt="Dr. STONE"
              class="makeStyles-thumbnail-21"
              src="https://mangadex.org/images/manga/20882.jpg?1585634146"
            />
          </a>
          <div
            class="MuiGrid-root makeStyles-infoGrid-25 MuiGrid-container MuiGrid-direction-xs-column"
          >
            <table
              class="makeStyles-infoTable-27"
            >
              <tbody>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Latest release:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-26 MuiTypography-body1"
                      title="Sun, 05 Jul 2020 16:00:00 GMT"
                    >
                      Jul 5th 2020, 16:00 - 1 day ago
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Estimated release interval:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-26 MuiTypography-body1"
                    >
                      7 days 0 hours
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Estimated next release:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-26 MuiTypography-body1"
                    >
                      Jul 12th 2020, 16:00
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Latest chapter:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-26 MuiTypography-body1"
                    >
                      157
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div>
          <div
            class="MuiToolbar-root MuiToolbar-regular makeStyles-root-35 MuiToolbar-gutters"
          >
            <button
              aria-label="Add item"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              title="Add"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <h6
              aria-label="Table title"
              class="MuiTypography-root makeStyles-title-36 MuiTypography-h6"
            >
              Forced checks
            </h6>
          </div>
          <table
            class="MuiTable-root"
            role="table"
          >
            <thead
              class="MuiTableHead-root makeStyles-tableHead-33"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head"
                role="row"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                />
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                >
                  Service name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                >
                  Service id
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                />
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root"
              role="rowgroup"
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Manga admin page should render correctly should render correctly without data 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
    >
      <div
        class="MuiPaper-root makeStyles-paper-5 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="makeStyles-titleBar-2"
        >
          <h4
            class="MuiTypography-root makeStyles-title-1 MuiTypography-h4"
          />
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            title="Go back"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 9l1.42 1.42L8.83 14H18V4h2v12H8.83l3.59 3.58L11 21l-6-6 6-6z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="makeStyles-details-4"
        >
          <a
            rel="noreferrer noopener"
            target="_blank"
          >
            <img
              class="makeStyles-thumbnail-3"
            />
          </a>
          <div
            class="MuiGrid-root makeStyles-infoGrid-7 MuiGrid-container MuiGrid-direction-xs-column"
          >
            <table
              class="makeStyles-infoTable-9"
            >
              <tbody>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Latest release:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-8 MuiTypography-body1"
                      title="Unknown"
                    >
                      Unknown
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Estimated release interval:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-8 MuiTypography-body1"
                    >
                      Unknown
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Estimated next release:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-8 MuiTypography-body1"
                    >
                      Unknown
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Latest chapter:
                    </p>
                  </td>
                  <td>
                    <p
                      class="MuiTypography-root makeStyles-detailText-8 MuiTypography-body1"
                    >
                      Unknown
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div>
          <div
            class="MuiToolbar-root MuiToolbar-regular makeStyles-root-17 MuiToolbar-gutters"
          >
            <button
              aria-label="Add item"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              title="Add"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <h6
              aria-label="Table title"
              class="MuiTypography-root makeStyles-title-18 MuiTypography-h6"
            >
              Forced checks
            </h6>
          </div>
          <table
            class="MuiTable-root"
            role="table"
          >
            <thead
              class="MuiTableHead-root makeStyles-tableHead-15"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head"
                role="row"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                />
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                >
                  Service name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                >
                  Service id
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                />
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root"
              role="rowgroup"
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;
